Documentation/troubleshooting.md

```diff
@@ -1,4 +1,84 @@
+ * `Error adding route...`
+ * `Add L2 failed`
+ * `Failed to set up IP Masquerade`
+ * `Error registering network: operation not permitted`
+
+...then confirm that the user running `flannel` has the right permissions (or trying running with `sudo`)
+
+## Performance
+
+### Control plane
+flannel is known to scale to a very large number of hosts. If you see a delay in being able to contact the pods on a newly created host then that can indicate control plane problems. flannel doesn't need much CPU or RAM but the first thing to check would be that it has adaquate resources available. flannel is also reliant on the performance of the datastore, either etcd or the Kubernetes API server. Check that they are performing well.
+
+### Data plane
+flannel relies on the underlying network so that's the first thing to check if you're seeing poor data plane performance.
+
+There are two flannel specific choices that can have a big impact on performance
+1) The backend being used. For examplem, if encapsulation is being used then `vxlan` will always perform better than `udp`. Or even better, avoid encapsulation for the maximum data plane performance.
+2) The MTU can have a large impact. To achieve maximum raw bandwidth, a network supporting a large MTU should be used. flannel writes an MTU setting to the `subnet.env` file. This file is read by either the Docker daemon or the CNI flannel plugin which does the networking for individual containers. To roubleshoot, the first step should be to ensure that the network interface that flannel is using has the right MTU. Then check that the correct MTU is written to the `subnet.env`. Finally, check that the containers have the correct MTU on their virtual ethernet device.
+
+
+## Firewalls
 When using `udp` backend, flannel uses UDP port 8285 for sending encapsulated packets.
+
 When using `vxlan` backend, kernel uses UDP port 8472 for sending encapsulated packets.
+
 Make sure that your firewall rules allow this traffic for all hosts participating in the overlay network.
+
+# Kubernetes Specific
+The flannel kube subnet manager relies on the fact that each node already has a `podCIDR` defined.
+
+You can check the podCidr for your nodes with one of the following two commands
+* `kubectl get nodes -o jsonpath='{.items[*].spec.podCIDR}'`
+* `kubectl get nodes -o template --template={{.spec.podCIDR}}`
+
+If your nodes do not have a podCIDR, then either use the --pod-cidr kubelet command-line option or the --allocate-node-cidrs=true --cluster-cidr=<cidr> controller-manager command-line options.
+
+If `kubeadm` is being used then pass `--pod-network-cidr=10.244.0.0/16` to `kubeadm init`
+
+It's possible to manually set the `podCIDR` for each node but this manual approach isn't recommended.
+* `kubectl patch node <NODE_NAME> -p '{"spec":{"podCIDR":"<SUBNET>"}}'`
```

Comment:
I meant, if it's not recommended, you might want to remove the entire paragraph, and just not tell people how to do it.